subprojects/core/src/main/java/org/gradle/execution/plan/ExecutionNodeAccessHierarchy.java

```diff
@@ -69,13 +69,30 @@ public void visitChildren(Iterable<NodeAccess> values, Supplier<String> relative
             @Override
             public void visitChildren(Iterable<NodeAccess> values, Supplier<String> relativePathSupplier) {
                 String relativePathFromLocation = relativePathSupplier.get();
-                if (filter.isSatisfiedBy(new ReadOnlyFileTreeElement(new File(location + "/" + relativePathFromLocation), relativePathFromLocation, stat))) {
+                if (relativePathMatchesSpec(filter, new File(location, relativePathFromLocation), relativePathFromLocation)) {
                     values.forEach(this::addNode);
                 }
             }
         });
     }
 
+    private boolean relativePathMatchesSpec(Spec<FileTreeElement> filter, File element, String relativePathString) {
+        boolean elementIsFile = element.isFile();
+        RelativePath relativePath = RelativePath.parse(elementIsFile, relativePathString);
+        if (!filter.isSatisfiedBy(new ReadOnlyFileTreeElement(element, relativePath, stat))) {
+            return false;
+        }
+        // All parent paths need to match the spec as well, since this is how we implement the file system walking for file tree.
+        File parentFile = element.getParentFile();
+        RelativePath parentRelativePath = relativePath.getParent();
+        for (; parentRelativePath != null && parentRelativePath != RelativePath.EMPTY_ROOT; parentRelativePath = parentRelativePath.getParent(), parentFile = parentFile.getParentFile()) {
```

Comment:
I find this relatively hard to read.

Reply:
I changed this to a while loop. Is it easier to read now?